### DISCUSSION TITLE
pixmap must be grayscale or rgb to write as png

### DISCUSSION THREAD

**Discussion by devpro9219** [pymupdf/PyMuPDF#1880] [Q&A] [answered]
Hi. I've tried to use this guide.
https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-imga.py

When I try to read all images from the pdf, I got this error. 
```
mupdf: pixmap must be grayscale or rgb to write as png
Traceback (most recent call last):
  File "image.py", line 97, in <module>
    imgdata = pix.getPNGData()
  File "/home/qwe/ocr-env/lib/python3.6/site-packages/fitz/fitz.py", line 4170, in getPNGData
    barray = self._getImageData(1)
  File "/home/qwe/ocr-env/lib/python3.6/site-packages/fitz/fitz.py", line 4151, in _getImageData
    return _fitz.Pixmap__getImageData(self, format)
RuntimeError: pixmap must be grayscale or rgb to write as png
```
Do you have any idea or can you help me? 
I attached the PDF file what I've tried. 
Thank you. 
[BCBSMI EOB.pdf](https://github.com/pymupdf/PyMuPDF/files/4349984/BCBSMI.EOB.pdf)

**Comment by JorjMcKie** [comment]
There obviously are images in the pdf with more than 3 color components. You must either store those in some CMYK image Format (png won’t work) or convert it to RGB first. I’m underway currently, so please have a look at examples extract-imga.py, where this is done, too.

Von meinem iPhone gesendet

Am 18.03.2020 um 12:02 schrieb IT Engineer. <notifications@github.com>:

﻿

Hi. I've tried to use this guide.
https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-imga.py

When I try to read all images from the pdf, I got this error.

mupdf: pixmap must be grayscale or rgb to write as png
Traceback (most recent call last):
  File "image.py", line 97, in <module>
    imgdata = pix.getPNGData()
  File "/home/qwe/ocr-env/lib/python3.6/site-packages/fitz/fitz.py", line 4170, in getPNGData
    barray = self._getImageData(1)
  File "/home/qwe/ocr-env/lib/python3.6/site-packages/fitz/fitz.py", line 4151, in _getImageData
    return _fitz.Pixmap__getImageData(self, format)
RuntimeError: pixmap must be grayscale or rgb to write as png


Do you have any idea or can you help me?
I attached the PDF file what I've tried.
Thank you.
BCBSMI EOB.pdf<https://github.com/pymupdf/PyMuPDF/files/4349984/BCBSMI.EOB.pdf>

**Comment by devpro9219** [comment]
Hi, I've already run https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-imga.py and It shows the above issues. 

Please give me advise. 
Thanks

**Comment by JorjMcKie** [comment]
Ok, I am back home now. Let me check your PDF.

**Comment by devpro9219** [comment]
Thank you.

**Comment by JorjMcKie** [comment]
Aha, resolved the issue:

The latest PyMuPDF also accepts the ICC color system, therefore corlorspaces may be presented which do have the right number of color components but still are neither DeviceGRAY, nor DeviceRGB. This required an adjustment of ``extract-imga.py``. Here is an update:
[extract-imga.zip](https://github.com/pymupdf/PyMuPDF/files/4350984/extract-imga.zip)

**Comment by devpro9219** [comment]
Thank you. It seems like it works. 
However, I have one more question, It seems like it doesn't handle the mask area from the pdf. 
Also, some images didn't extracted. Do you have any sample to handle those things based on your library? 
Thank you. 
![Screenshot_2](https://user-images.githubusercontent.com/36050944/77000970-90169580-692f-11ea-82ba-a73d6867cec7.png)
![Uploading Screenshot_3.png…]()

**Comment by devpro9219** [comment]
This is the part which can't detect with this library. 
Please give me the advise. 
Thank you. 
![Screenshot_3](https://user-images.githubusercontent.com/36050944/77001081-b9372600-692f-11ea-86f4-85597eea91c6.png)

**Comment by JorjMcKie** [comment]
Your PDF is a complex example! You almost have to rewrite a PDF viewer for a full analysis. Here are my findings:

1. The images extracted by the script are the images which really exist - there are no other ones.
2. The red rectangle in screenshot 3 is **not** an image but a so-called XObject (a page embedded from another PDF). I have added a script which is able to extract those things ...
3. Other apparently graphical elements are **no** images but drawings generated by the PDF-internal mini-language (similar to PostScript). The red frame in your last post is such a **non-image**. There is no way to extract these things (except, as mentioned, by writing a PDF viewer-like script)

Here is a more advanced text extraction script, which should extract the text in the correct reading sequence:
[doubles.zip](https://github.com/pymupdf/PyMuPDF/files/4351360/doubles.zip)

Here is a script to extract XObjects (like the red box on page 2):
[xobj-extract.zip](https://github.com/pymupdf/PyMuPDF/files/4351361/xobj-extract.zip)

**Comment by devpro9219** [comment]
Hi. Thanks for your advise. However, I need to get exactly same image with the current PDF. 
However, this is my question. 
- How to get the image position and width and height on the PDF? 
For example, the attachment 2, the image is shown the part of image on the pdf. 
However, I get the full document image(attached). 
And the color of the computer image is not same as the PDF. 

There is any way to detect the mask or effect for each images?
Thanks
![img-46](https://user-images.githubusercontent.com/36050944/77009683-b68ffd00-693e-11ea-89fa-1c3293ae8029.png)

![img-41](https://user-images.githubusercontent.com/36050944/77009656-aa0ba480-693e-11ea-8f14-83eac091ce62.png)

**Comment by devpro9219** [comment]
I've also read fully document, however, I didn't catch how I can get the x,y,width,height information about the image. 
Is it possible? 
Thanks!

**Comment by JorjMcKie** [comment]
I don't know your full motivation behind all this. But here are a few hints that may help:

```python
doc = fitz.open("BCB...")
page=doc[1]  # page 2
imglist = doc.getPageImageList(1, True)  # full image list of that page
bbox = page.getImageBbox(imglist[0])  # this is img-46.png
# just to demonstrate we do have it:
page.addRectAnnot(bbox)  # gives this:
```
![grafik](https://user-images.githubusercontent.com/8290722/77046481-71e57f80-6999-11ea-811e-d60b87c92b47.png)

When we extract the images, the mask is automatically detected and applied! My own extraction produced this. Not 100% the same colors, but good. The difference is probably caused by the conversion to RGB.

![img-46](https://user-images.githubusercontent.com/8290722/77046830-0fd94a00-699a-11ea-8d5d-a9e5ee268bb9.png)

You can also try to **not** convert to RGB in that script. Use an image format which supports CMYK like PAM or Photoshop image (PSD):

```python
...
if pix.colorspace.name not in (fitz.csGRAY.name, fitz.csRGB.name):
    pix.writeImage("xxxx.pam")
```

But in your case this does not work either. If you use the MuPDF command line tool ``mutool extract <infile.pdf>`` (which extracts images **and** fonts), the results are even worse.

So I guess that is what you can get from me ...

> Note: you should use PyMuPDF v1.16.13! It contains ICC color support and your problem image needs this for optimal rendering.

**Comment by JorjMcKie** [comment]
@devpro9219 - Assuming your questions were answered.
Please do not hesitate to re-open or open new issues.

**Comment by aleem75321** [comment]
Hi, this script fo working fine for me but extract the image in Grey in my pdf file all image in CMYK  formate can you help how solve this

**Comment by JorjMcKie** [comment, marked as answer]
> Hi, this script fo working fine for me but extract the image in Grey in my pdf file all image in CMYK formate can you help how solve this

You must find a file format supporting CMYK. There are a selected few directly supported by `Pixmap.save()`. If none suits you, check the Pillow documentation for one such and then use `Pixmap.pil_save(...)` with the right parameters - again, please consult Pillow docu for choosing the right parameters in place of `...`.

**Reply by JorjMcKie** [reply]
You also need not take a pixmap as intermediate if you have the image file or binary via `doc.extract_image(xref)`.
In that case simply use Pillow directly.
